File: src/modifiers/rotationEffect.ts
import { createModifier } from './createModifier';

/// Anchor point for a transform, expressed as a named `UnitPoint`.
export type TransformAnchor =
  | 'center'
  | 'top'
  | 'bottom'
  | 'leading'
  | 'trailing'
  | 'topLeading'
  | 'topTrailing'
  | 'bottomLeading'
  | 'bottomTrailing';

export interface RotationEffectParams {
  /// Rotation angle in degrees (positive is clockwise).
  degrees: number;
  /// Point the view rotates about. Defaults to `'center'`.
  anchor?: TransformAnchor;
}

/// SwiftUI `.rotationEffect(_:anchor:)`. Rotates the view in 2D by the
/// given angle (degrees) about `anchor` without affecting layout.
export function rotationEffect(
  params: RotationEffectParams
): ReturnType<typeof createModifier>;
export function rotationEffect(
  degrees: number,
  anchor?: TransformAnchor
): ReturnType<typeof createModifier>;
export function rotationEffect(
  a: RotationEffectParams | number,
  b?: TransformAnchor
) {
  if (typeof a === 'number') {
    return createModifier('rotationEffect', { degrees: a, anchor: b });
  }
  return createModifier('rotationEffect', a);
}

File: apple/Sources/ReactNativeWatchOS/Modifiers/RotationEffect.swift
import SwiftUI

/// Maps a named anchor string to a `UnitPoint` for transform modifiers.
/// Kept namespaced (rather than a free function) to avoid colliding with
/// the gradient views' `parseUnitPoint`. Defaults to `.center`.
enum RNWTransformAnchor {
    static func unitPoint(_ s: String?) -> UnitPoint {
        switch s {
        case "top":            return .top
        case "bottom":         return .bottom
        case "leading":        return .leading
        case "trailing":       return .trailing
        case "topLeading":     return .topLeading
        case "topTrailing":    return .topTrailing
        case "bottomLeading":  return .bottomLeading
        case "bottomTrailing": return .bottomTrailing
        default:               return .center
        }
    }
}

/// SwiftUI `.rotationEffect(_:anchor:)`. Rotates the view in 2D by
/// `degrees` about `anchor` without affecting its layout.
enum RNWRotationEffectModifier {
    @MainActor
    static func register(into r: RNWModifierRegistry) {
        r.register("rotationEffect") { view, params, _ in
            let degrees = params.double("degrees") ?? 0
            let anchor = RNWTransformAnchor.unitPoint(params.string("anchor"))
            return AnyView(view.rotationEffect(.degrees(degrees), anchor: anchor))
        }
    }
}
